java/server/src/org/openqa/selenium/grid/sessionqueue/local/LocalNewSessionQueue.java

```diff
@@ -55,10 +56,11 @@
 
   private static final Logger LOG = Logger.getLogger(LocalNewSessionQueue.class.getName());
   private final EventBus bus;
-  private final Deque<HttpRequest> sessionRequests = new ConcurrentLinkedDeque<>();
+  private final Deque<RequestId> requestIdQueue = new ConcurrentLinkedDeque<>();
+  private final Map<RequestId, HttpRequest> sessionRequests = new ConcurrentHashMap<>();
```

Comment:
Would it make sense to put `RequestId` and `HttpRequest` in a new structure/class, and then only keep `Deque`?

I am not sure if things could get messy by managing two concurrent arrays that share the same information.

Reply:
That is a fair concern. LocalNewSessionQueue is using a write lock before making any changes to both the data structures.  So anytime it holds the lock, only it will be able to update the two. The write lock is used for all the operations. This mapping is needed to poll the request based on the id. If the Distributor would call `remove(httpRequest)`, then we wouldn't need it but that is not efficient.

Reply:
Since we require deque functionality and there is no threadsafe version of LinkedHashMap, another data structure was required to store the mapping of id to request. I was also concerned that it might get difficult with two concurrent data structures, hence was relying on locks.  I will attempt to see if this can be changed to use just one data structure in some way and have the deque functionality.

Reply:
Do you think it might be a good idea to use` Collections.synchronizedMap(new LinkedHashMap())` and then perform operation in a thread-safe manner?

Reply:
The `Deque` is used for ordering, and the `Map` is used to store the actual request, right? I think our expected usage pattern is simply to take the item off the head of the `Deque` and just use that? With a very limited number of random removals of items, or other traversals looking for matches? If so, we might just want a holding data structure with two fields (`requestId, request`) and store _that_ in the `Deque`

Reply:
This change is related to https://github.com/SeleniumHQ/selenium/pull/8747#discussion_r497621655 . So the idea is to pass the request-id from the Distributor to remove the request from the queue instead of passing the request itself. Would it be okay performance-wise to traverse every time Distributor calls `remove(requestId)` ? Especially when all slots are occupied and we add it to the front of the `Deque` and retry ?

Reply:
I think this is ok because if a remove happens often it is a sign that the user has an underpowered Grid with many new session requests. 

For example, a Grid with two nodes and 2 slots on each node won't be able to cope with a test suite sending tests in 20 threads. That is a not well planned Grid. 

Since we are having observability as a core part of the new Grid, we could redirect users to look at that and figure out how many removals are happening in their tests.

Summarizing, `remove(requestId)` should not happen that much in a well planned Grid.

Reply:
Thank you so much for the detailed explanation. I get the idea clearly now. Good point regarding the observability bit, we have tracing for remove request. It will provide good insight.